### DISCUSSION TITLE
Hooks support

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#24893] [Suggest an Idea]
### Describe the proposed change(s).

Add possibility for "hooks", similar to postUpgradeTasks but simpler because they don't directly commit anything.

It could be like this nested structure:

```json
{
  "hooks": {
    "branch": {
      "create": [],
      "update": []
    },
    "pr": {
      "create": [],
      "update": []
    }
  }
}
```

It would need a corresponding `allowedHooks` array in global config defining patterns which are allowed.

### Describe why we need/want these change(s).

People might want to run code after PR creation, update, merging, closing, etc.

**Comment by stephen-dahl** [comment]
I would like a hook for after renovate clones a project. it would be a good escape hatch for when a repo needs some setup before renovate does its thing and we haven't added support yet.

In my case I waiting on #24460 but if I could add a post clone hook that runs `echo "myconfig with pipeline tokens" > bunfig.toml` I could work around until that issue is resolved.

**Reply by jamietanna** [reply]
https://github.com/renovatebot/renovate/discussions/40237 is another case where it would be useful to have a post-clone option - see https://github.com/renovatebot/renovate/discussions/40237#discussioncomment-15381604 for some thoughts on allowlisting of these commands

**Reply by staktrace** [reply]
I would also like a hook for post-clone operations. Our use case is a bazel repo where the MODULE.bazel file uses python constructs to unroll an array into maven.install/maven.artifact calls. Currently renovate does not handle this. We could write a pre-processor that makes the file compatible with Renovate's parser but currently there is no hook to run the preprocessor after cloning and before Renovate gets to work on the repo.

This discussion is about a general hooks support mechanism which is great, but would you be willing to take a shorter-term fix that just adds a preUpdate or similar hook?

**Reply by felipecrs** [reply]
I actually have a working implementation of `postInitTasks` (https://github.com/felipecrs/renovate/commit/04114a4593e1e59497d1644676ec75497cd65b87), but I don't think we should introduce it just to be deprecated later in favor of proper hooks.

**Comment by Ramshield** [comment]
I'd like this. When I update a Docker-compose file (with a minor version) I want it to do a restart so it uses the new version.
